src/Grpc.Net.Client/Balancer/Internal/SocketConnectivitySubchannelTransport.cs

```diff
@@ -232,37 +250,40 @@ private void OnCheckSocketConnection(object? state)
-                        // then the socket can't be used by SocketsHttpHandler, but the socket can get recreated when it is eventually used.
+                        // Poll socket to check if it can be read from. Unfortunatly this requires reading pending data.
+                        // The server might send data, e.g. HTTP/2 SETTINGS frame, so we need to read and cache it.
                         //
                         // Available data needs to be read now because the only way to determine whether the connection is closed is to
                         // get the results of polling after available data is received.
-                        //
-                        // If a socket with data is closed immediately then that might happen continuously.
-                        // Chill, leave the socket alone, and be happy that it can be read from. We can create a new one later.
+                        bool hasReadData;
                         do
                         {
-                            closeSocket = IsSocketInBadState(socket, socketAddress, allowAvailableReadBytes: true);
+                            closeSocket = IsSocketInBadState(socket, socketAddress);
                             var available = socket.Available;
                             if (available > 0)
                             {
-                                SocketConnectivitySubchannelTransportLog.SocketReceivingAvailable(_logger, _subchannel.Id, socketAddress, available);
-
-                                var buffer = ArrayPool<byte>.Shared.Rent(minimumLength: available);
-                                try
-                                {
-                                    // Data is already available so this won't block.
-                                    var readCount = socket.Receive(buffer);
-                                    _initialSocketDirtyBytesReadCount += readCount;
-                                }
-                                finally
+                                hasReadData = true;
+                                var serverDataAvailable = CalculateInitialSocketDataLength(_initialSocketData) + available;
+                                if (serverDataAvailable > MaximumInitialSocketDataLength)
                                 {
-                                    ArrayPool<byte>.Shared.Return(buffer);
+                                    throw new InvalidOperationException($"The server sent {serverDataAvailable} bytes to the client before a connection was established. This exceeds maximum data allow.");
                                 }
+
+                                SocketConnectivitySubchannelTransportLog.SocketReceivingAvailable(_logger, _subchannel.Id, socketAddress, available);
+
+                                // Data is already available so this won't block.
+                                var buffer = new byte[available];
```

Comment:
for consideration: should this use a pool (`ArrayPool<byte>.Shared` or similar) and return chunks when they've been processed later

Reply:
I didn't do that right now because I expect the amount of data will always be small (SETTINGS frame) and it only happens when the connection is started. If it is highlighted as a perf issue then it can be revisited.